yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,119 @@ namespace flowable {
 template <typename T>
 class Subscriber : public virtual Refcounted, public yarpl::enable_get_ref {
  public:
+  virtual void onSubscribe(Reference<Subscription>) = 0;
+  virtual void onComplete() = 0;
+  virtual void onError(folly::exception_wrapper) = 0;
+  virtual void onNext(T) = 0;
+};
+
+// codemod all things that inherit from Subscriber to inherit from
+// InternalSubscriber
+template <typename T>
+class InternalSubscriber : public Subscriber<T> {
+ public:
+  // Note: If any of the following methods is overridden in a subclass, the new
+  // methods SHOULD ensure that these are invoked as well.
+  void onSubscribe(Reference<Subscription> subscription) override {
+    DCHECK(subscription);
+    CHECK(!subscription_);
+    subscription_ = std::move(subscription);
+  }
+
+  // No further calls to the subscription after this method is invoked.
+  void onComplete() override {
+    DCHECK(subscription_) << "Calling onComplete() without a subscription";
+    subscription_.reset();
```

Comment:
I would ask one question about this. What if we think in reverse, how? Instead of keeping `subscription_` object in the parent class and providing a function to the child class to access this object, can we have a function, which is virtual that the parent class can call and get access to the `subscription_` data member in the child class?

So parent will not keep and reference to the `subscription_` object, but still can access it through that virtual function call. 

In that case, the parent will never need to call `subscription_.reset()` at all, as it already doesn't hold any reference to the `subscription_` permanently.